File: tests/test_tcc.c
/**
 * @file test_tcc.c
 * @brief Test suite for Thai Character Cluster (TCC)
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "tcc.h"

int main() {
    printf("=== CThaiNLP TCC Test Suite ===\n");
    
    int passed = 0;
    int total = 0;
    
    /* Test 1: basic word segmentation */
    {
        total++;
        int count = 0;
        char** tokens = tcc_segment("ฉันไปโรงเรียน", &count);
        /* Expected: ['ฉั', 'น', 'ไป', 'โรง', 'เรี', 'ยน'] */
        if (tokens && count == 6 &&
            strcmp(tokens[0], "ฉั") == 0 &&
            strcmp(tokens[1], "น") == 0 &&
            strcmp(tokens[2], "ไป") == 0 &&
            strcmp(tokens[3], "โรง") == 0 &&
            strcmp(tokens[4], "เรี") == 0 &&
            strcmp(tokens[5], "ยน") == 0) {
            printf("[Test %d] TCC basic word segmentation: PASS\n", total);
            passed++;
        } else {
            printf("[Test %d] TCC basic word segmentation: FAIL (count=%d)\n", total, count);
        }
        tcc_free_result(tokens, count);
    }
    
    /* Test 2: empty string */
    {
        total++;
        int count = 0;
        char** tokens = tcc_segment("", &count);
        if (!tokens && count == 0) {
            printf("[Test %d] TCC empty string: PASS\n", total);
            passed++;
        } else {
            printf("[Test %d] TCC empty string: FAIL\n", total);
        }
        tcc_free_result(tokens, count);
    }
    
    /* Test 3: tcc_pos */
    {
        total++;
        int* positions = NULL;
        int count = tcc_pos("ฉันไปโรงเรียน", &positions);
        if (positions && count == 6) {
            printf("[Test %d] TCC positions: PASS (count=%d)\n", total, count);
            passed++;
        } else {
            printf("[Test %d] TCC positions: FAIL\n", total);
        }
        free(positions);
    }
    
    printf("\n=== TCC Test Summary: %d/%d passed ===\n", passed, total);
    return (passed == total) ? 0 : 1;
}

File: tests/python/test_tcc.py
#!/usr/bin/env python3
"""
Test suite for CThaiNLP TCC module
"""

import unittest
import os
import sys

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../..')))

from cthainlp import tcc
from cthainlp.tokenize import tcc as tok_tcc
from cthainlp.tokenize import subword_tokenize


class TestTCC(unittest.TestCase):
    """Test Thai Character Cluster functions"""

    def test_segment(self):
        self.assertEqual(tcc.segment(None), [])
        self.assertEqual(tcc.segment(""), [])
        res = tcc.segment("ฉันไปโรงเรียน")
        self.assertIsInstance(res, list)
        self.assertEqual("".join(res), "ฉันไปโรงเรียน")
        self.assertEqual(tcc.segment("ฉันไปโรงเรียน"), ["ฉั", "น", "ไป", "โรง", "เรี", "ยน"])

    def test_tcc_generator(self):
        gen = tcc.tcc("สวัสดี")
        items = list(gen)
        self.assertGreater(len(items), 0)
        self.assertEqual("".join(items), "สวัสดี")

    def test_tcc_pos(self):
        self.assertEqual(tcc.tcc_pos(None), set())
        self.assertEqual(tcc.tcc_pos(""), set())
        pos = tcc.tcc_pos("ฉันไปโรงเรียน")
        self.assertIsInstance(pos, set)
        self.assertIn(len("ฉันไปโรงเรียน"), pos)

    def test_module_aliases(self):
        self.assertEqual(tcc.segment("ไป"), tok_tcc.segment("ไป"))
        self.assertEqual(tcc.tcc_pos("ไป"), tok_tcc.tcc_pos("ไป"))

    def test_subword_tokenize(self):
        res = subword_tokenize("ฉันไปโรงเรียน", engine="tcc")
        self.assertEqual(res, ["ฉั", "น", "ไป", "โรง", "เรี", "ยน"])
        
        # Test keep_whitespace
        text = "ไป โรงเรียน"
        self.assertIn(" ", subword_tokenize(text, keep_whitespace=True))
        self.assertNotIn(" ", subword_tokenize(text, keep_whitespace=False))

        with self.assertRaises(ValueError):
            subword_tokenize("ฉัน", engine="invalid_engine")


if __name__ == "__main__":
    unittest.main()
